Lock multiple resources using labels and quantities within the Pipeline lock step

Aeon512 opened the pull request:
As a result of [JENKINS-34268](https://issues.jenkins-ci.org/browse/JENKINS-34268) and [JENKINS-34273](https://issues.jenkins-ci.org/browse/JENKINS-34273) I tried to extend the current Pipeline lock step implementation with an additional optional argument `label: "LabelName"` to specify that resources assigned to this label shall be requested as well as an additional optional argument `quantity: 3` to request 3 such resources. (If left out, quantity defaults to 1.)

The locking and unlocking mechanisms had to be largely modified in order to deal with several resources in parallel (the current pipeline implementation can only handle a single resource within one step). One major modification was to move the list of queued context out of the actual `LockableResource` (since we are now dealing with several resources at once) and move this queue to the `LockableResourcesManager` instead.

Example Usage:

```
lock(label: "MyLabel", quantity: 2) {
    // body
}
```

Since this is my first Jenkins Plugin contribution (and the first time writing actual Java code) I'm open for any sorts of feedback, review comments or similar. The same holds actually for GitHub, so if there is anything I can improve doing forks, commits and pull requests, please let me know :)

As of today, 2016-04-22, this fork is up-to-date with the base repository, all unit-tests can be executed, and additional unit-tests for the newly added functionality (labels & quantities) have been added. The plugin worked also fine in my manual Jenkins tests.

amuniz commented on src/main/java/org/jenkins/plugins/lockableresources/LockStep.java:
```diff
@@ -11,25 +11,45 @@
 	}
 
+	@DataBoundSetter
+	public void setResource(@Nullable String resource) {
```
More clear if you add `@CheckForNull` in the field rather that using `@Nullable` here.

amuniz commented on src/main/java/org/jenkins/plugins/lockableresources/LockStep.java:
```diff
@@ -11,25 +11,45 @@
+	}
+
+	@DataBoundSetter
+	public void setLabel(@Nullable String label) {
```
Ditto. And just assign it without any check.

amuniz commented on src/main/java/org/jenkins/plugins/lockableresources/LockStep.java:
```diff
@@ -11,25 +11,45 @@
+	@DataBoundSetter
+	public void setLabel(@Nullable String label) {
+		if (label != null && !label.isEmpty())
+		{
```
Use inlined `{` please:

```
if (...) {
    ...
} else {
    ...
}
```
I'm preparing contributing guidelines to define this.

amuniz commented on src/main/java/org/jenkins/plugins/lockableresources/LockableResource.java:
```diff
@@ -65,12 +65,6 @@
-	 * Only used when this lockable resource is tried to be locked by {@link LockStep},
-	 * otherwise (freestyle builds) regular Jenkins queue is used.
-	 */
-	private List<StepContext> queuedContexts = new ArrayList<StepContext>();
```
:-1: this is backward incompatible at configuration file level

Aeon512 replied:
However, once you have several resources, where would you store the queuedContexts? There is no longer one single resource (as in your implementation) to handle this.

amuniz commented on src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java:
```diff
@@ -41,6 +44,13 @@
+	 * Only used when this lockable resource is tried to be locked by {@link LockStep},
+	 * otherwise (freestyle builds) regular Jenkins queue is used.
+	 */
+	private List<QueuedContextStruct> queuedContexts = new ArrayList<QueuedContextStruct>();
```
Moving this field here makes it hidden to static serialized state (needed for restart-ability).

Aeon512 replied:
As far as my limited understanding goes, this field is serialized in the configuration file once `save()` is called. Since whenever something is added or removed to this list, `save()` is called, this should survive a restart.

amuniz replied:
Yes, you are right, I missed the list is now of `QueuedContextStruct` instead of `LockableResource`, and that's why the restart tests are passing :)

amuniz commented on src/main/java/org/jenkins/plugins/lockableresources/LockableResourcesManager.java:
```diff
@@ -210,74 +220,176 @@ private boolean checkCurrentResourcesStatus(List<LockableResource> selected,
-						}
+	public synchronized void unlock(@Nullable List<LockableResource> resourcesToUnLock,
+			Run<?, ?> build, @Nullable StepContext context,
+			@Nullable List<String> resourceNamesToUnLock, boolean inversePrecedence) {
```
I think this new parameter (`resourceNamesToUnLock`) is going to be mutually exclusive with the original one (`resourcesToUnLock`). A new method should be created to manage this new override.

amuniz commented:
After a quick review (as it would need a deeper one), it seems that the static state is not persisted now (queued step contexts). Why the restart tests are passing? I don't know, probably something to fix in tests.

Aeon512 commented:
@amuniz Thanks for the quick review. I have addressed most of your points except the still open issue regarding queued step contexts. Your comment regarding static state is beyond my understanding.

As mentioned in the comment above, I think the queued step context is correctly stored by issuing the `save()` call. The problem with storing the queued step contexts in individual resources is mostly, that by using labels you don't know apriori which resource will be finally used within the lock as their availability might change. Hence, you would have to store the queue within all possible resource and maintain these different lists which yield to a quite complicated solution imho.

Any ideas how to progress forward here?

amuniz commented:
@Aeon512 I need to spend some time doing a deeper review. But at least, you will need to add a migration process to keep backward compatibility with previous installations. That migration would need to move all waiting contexts in the old structure to the new one. For that purpose the old field `LockableResources#queuedContexts` needs to be kept (not removed but `@Deprecated`). The migration process can be implemented by using `@InitMilestone`. Perhaps you can find some similar example [here](https://github.com/search?p=2&q=org%3Ajenkinsci+InitMilestone&type=Code).

Aeon512 commented:
@amuniz Thanks a lot for your work so far. I would really appreciate if you could take the time for doing a deeper review.

I have restored the old field `LockableResource#queuedContexts` as well as a new getter to obtain its data `LockableResource#getQueuedContexts`. Reharding the migration process I have tried to implement some code within `BackwardCompatibility.java` but honestly speaking that's only a shot from the hip since I don't really know how to verify / test this scenario.

Aeon512 commented:
Any news how this will progress?

amuniz commented:
@Aeon512 I didn't have time yet :cry: I'm a bit concerned about how this PR impacts freestyle compatibility and how the migration works, as we don't have too much automated tests for the former and none for the latter. So merging without manual testing is kind of dangerous (and they require time). Adding that tests is in my list, but feel free to add some if you feel good :smiley: 

Migration tests are implemented in other plugins by using `@LocalData`, [this is an example](https://github.com/jenkinsci/jgiven-plugin/blob/0403591e55f630b60d808458227b25b92ba8c524/src/test/java/org/jenkinsci/plugins/jgiven/CompatibilityTest.java#L16).

Aeon512 commented:
@amuniz ping

talcloudshare commented:
Any news? Specifically about JENKINS-34268?
Thanks

Aeon512 commented:
@talcloudshare Unfortunately not. Sadly, I haven't heard from @amuniz 
Working with a locally build version of the plugin for the moment.

amuniz commented:
Sorry, I didn't have time to review this, will do it as soon as I can.

amuniz commented on src/main/java/org/jenkins/plugins/lockableresources/BackwardCompatibility.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Initializer(after = InitMilestone.JOB_LOADED)
+    public static void beforeInitMilestonePluginsStarted() {
+        LOG.info("\n\n LOCKABLE RESOURCES COMPATABILITY INIT \n\n");
```
Better at `FINE` level.

amuniz commented on src/main/java/org/jenkins/plugins/lockableresources/BackwardCompatibility.java:
```diff
@@ -0,0 +1,46 @@
+					List<String> resourcesNames = new ArrayList<String>();
+					resourcesNames.add(resource.getName());
+					LockableResourcesStruct resourceHolder = new LockableResourcesStruct(resourcesNames, "", 0);
+					LockableResourcesManager.get().queueContext(queuedContext, resourceHolder, resource.getName());
```
Shouldn't be resource#queuedContexts cleared up after this migration? So it does not take effect anymore.

amuniz commented:
The snippet generator is allowing to create:

```
lock(inversePrecedence: true, label: 'label1', quantity: 3, resource: 'resource1') {
    // some block
}
```

Without any warning. As `label` and `resource` parameters are mutual exclusive a validation should be shown, and a WARNING log in runtime saying that `label` will be ignored as `resource` is specified (not sure if that's the current behaviour, but it should be).

Aeon512 commented:
@amuniz How can I mark parameters as mutual exclusive?

amuniz commented:
1) Form validation **and**
2) Exception throwing at runtime if both are set

amuniz commented on src/main/java/org/jenkins/plugins/lockableresources/LockStep.java:
```diff
@@ -11,25 +11,47 @@
-			throw new IllegalArgumentException("must specify resource");
-		}
-		this.resource = resource;
+	public LockStep() {
```
This is backward incompatible, but I fear there is no way to preserve it in this case.
Any call to to `lock` in the form `lock ('my-resource') { ... }` is going to fail.

@jglick any idea on how keep backward compatibility here?

amuniz replied:
This is preventing me to go ahead and merge. Unfortunately I don't see how keep backward compatibility here...

Aeon512 replied:
You are right that this breaks backward compatibility.
But only because the first pipeline implementation done by cloudbees only focused on the usage of resource names (and skipped the usage of labels and quantities all together).

One could argue that either
- the first pipeline implementation by cloudbees shouldn't have been merged at all, because it focused only on 33% and did not provide an upgrade path (the problem you are looking at now)
- that increasing the functionality for the pipeline implementation to the full feature set of the classic non-pipeline implementation is allowed to introduce breaking-changes since pipelines are still under heavy development.

Hence, I don't agree with your assumption that due to a incomplete (wrong) implementation done previously, we are not able to correct this implementation and do it right.
Sometimes a breaking change must happen. The sooner the better.

Vlatombe replied:
I may be missing something here but I don't think this has to be a breaking change. `resource` can be left in the constructor, and you can use `@DataBoundSetter` for the new fields. Both should work just fine.

Aeon512 commented:
Do you have an example for a form validation in the pipeline snippet generator?
Couldn't get this to work.

Aeon512 commented:
Anyone has an idea why the latest build got stuck? Didn't change much and locally everything works

amuniz commented:
> Do you have an example for a form validation in the pipeline snippet generator?

It's regular Jenkins (stapler) validation, [this is an example](https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/5aa2b7c7133dfda401d640683b09d3f960a02546/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/BitbucketSCMSource.java#L543).

Aeon512 commented:
@amuniz I have added the form validation.
However, I have no idea why the automatic build failed. Locally everything works fine.

RemcoTukker commented:
Looks great! One question though: looking through the PR, it seems there's no way of finding out which resources you actually locked when using a label (eg through a global variable), or did I just miss it?

I think it would be very useful, at least when your resources are physical things. How else will you decide which resource actually to use after you acquired a lock on it? (Any one that's free is not good enough, because it could be free while someone else has a lock on it..)

chancez commented:
I'd love to see this. I currently need a better way than the built-in throttle functionality to limit the number of builds in a multibranch pipeline so that our Jenkins executors aren't overwhelmed with builds from a single repo with lots of branches/PRs.

bmannix commented:
Is this PR still under development, or has it been abandoned?

Aeon512 commented:
From my side this PR was finished. It builds locally without any problems and works fine since then. 
Why the build on the Jenkins CI got stuck is unknown to me.
unfortunately Maintenance / Integration of Community PR seems to have a low priority at CloudBees...

amuniz commented:
Basically it was not merged because of the backward incompatible change, but @Vlatombe is pointing out that there isn't such backward incompatibility if the original `@DataboundConstructor` is kept and a `@DataboundSetter` is added for the new field.

The snippet generator is going to request for `resource` field always though, but I think it's a minor issue compared to the number of current calls the constructor change would break.

This is in my list, I can make and test the change by myself.

amuniz commented:
@Vlatombe I just checked the snippet generation is completely broken if we keep the original constructor.

@Aeon512 I'm trying to fix it (will send some commits on top of yours).

amuniz commented:
A follow up on this PR has been filed as #42 - it is ready for review and merge.
